### ISSUE TITLE
Question: Always need to clone a state property in reducer?

### ISSUE THREAD

**Issue by Mika83AC** [tur-nr/polymer-redux#70]
Hello again,

given follwing example:

```
// state has property 'list' which is an array().

const reducer = (state, action) => {
  switch (action.type) {
    case 'ADD_LIST_ITEM':
      state.list.push('new item');
      return Object.assign({}, state, { list: state.list });
  }
};
```

Using this, a polymer element property bound by "statePath: 'list'" doesn't recognize the change of state.list. 

When doing the following, everything works like desired:

```
const reducer = (state, action) => {
  switch (action.type) {
    case 'ADD_LIST_ITEM':
      var tmp = state.list.splice(0);
      tmp .push('new item');
      return Object.assign({}, state, { list: tmp });
  }
};
```

So does this mean, I've to copy/clone every property of the state in the reducer before changing it?

Regards,
Michael

**Comment by tur-nr** [comment]
Redux's state is `readonly`, so you can't mutate the array and return it. You must send a copy with the mutations otherwise you break the pattern. Same applies if you're reordering arrays.

Don't worry about memory, providing you're following both polymer-redux and redux's docs the browser will garbage collect old values.

If you need to hold reference to the array and apply mutations. You need to rethink why you're using redux and stick to polymer's helper methods.

**Comment by Mika83AC** [comment]
Memory and CPU usage was the point which let me think "copying multiple larger arrays for every single add/remove/update can't be right". Especially as my app should also run on mobile devices which are less less powerful as as desktop. But if this is "the pattern" and I want to use it, I have to use it. :) Larger applications can't live without a proper state management, and as Polymer itself it missing this completely, I'm happy that I've found PolymerRedux.

Thx for for your help again!

**Comment by tur-nr** [comment]
You don't _have_ to use this "pattern". Google wrote their own unidirectional, [UniFlow](https://google.github.io/uniflow-polymer/) with achieves the same goal just with a different state manager.

Whatever you use, unidirectional is the way forward. Always push your values down and use [Mediators](https://medium.com/@fyzaparviz/communication-in-polymer-1dc945a4d2ab) correctly. By correctly, I mean using `readOnly` on the properties that push state upwards.

**Comment by Mika83AC** [comment]
Yes, Google has UniFlow, which I used in my first approach for my app. But as it is not Polymer 2.x compatible and doesn't seem to be actively developed, I decided to give PolymerRedux a shot. Also PolymerRedux/Redux has much better debugging capabilities as UniFlow which should also be very helpful when developing larger apps.

So for now, if I can solve all my issues, I'll stick with PolymerRedux. :+1:
